### PR TITLE
Index mixin and add tests

### PR DESCRIPTION
**Pull request by khoahaha**
# Summary
- Index module mixin
- Inheritance specs did not cover the case where a class's superclass is defined after itself which eventually causes the mixin test to be broken. Fix is added for inheritance and its dedicated test will be in turn updated in another PR

# Inheritance Problem
Initially, processor `getconstant` creates a new ClassObject and labels it as module for any constants that the processor is at that time unable to locate.

Now assume that the constant is a superclass that gets referred to when its body definition not yet processed, it will be first wrongly saved as a module. When its `defineclass` pops out, it will be skipped at `@engine.define_class` since its definition was already there to begin with. If it itself were to be inherited from another class, this information would be completely left out.

The same goes to module, but it is defined correctly from the start and further it does not matter its `defineclass` gets skipped because module definition does not include extra information other than than that how it is previously created.

# Changes
- Register new processors: `include`, `extend` and `prepend`
- Add three new containers to hold different types of mixin-ed modules and their according index functions
- Add test
- Update the implementation of `instance_method_objects` and `class_method_objects`
- Update the implementation of engine's `define_class`

### CODE REVIEW

File: spec/definitions/mixin_spec.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+RSpec.describe 'index mixin-ed class' do
+  let(:class_name) { 'Man' }
+  let(:include_module) { 'UpOnTheSky' }
+  let(:extend_module) { 'UnderTheSea' }
+  let(:prepend_module) { 'OnTheGround' }
+  let(:code) do
+    <<-CODE
+      class Man
+        include UpOnTheSky
```
Comment (0x2c7) [inline spec/definitions/mixin_spec.rb]: This test is simple. Let's add some more tests that includes multiple prepends and includes; as well as normal methods to test out the behaviors.

File: lib/ruby_zen/class_object.rb [hunk 2]
```diff
@@ -35,6 +39,38 @@ def instance_method_objects(inherited = true)
       end
     end
 
+    def available_instance_methods
```
Comment (0x2c7) [inline lib/ruby_zen/class_object.rb]: Why don't we just replace `instance_method_objects` by this implementation?

**Review by 0x2c7 — CHANGES_REQUESTED** [review]
IMO, the ClassObject should be simple. It is the data storage and method searcher. The logic related to "re-define" guessing constant type should belong to higher level, in this case, Engine. The `define_class` should be responsible for this job.

```ruby
def define_class(class_name)
  class_name = class_name.to_s
  return @classes[class_name] if !block_given?

  class_object = yield
  if @classes[class_name]
    # Finalize class object type
    @classes[class_name].tap do |old_object|
       old_object.is_module = class_object.is_module
       old_object.namespace = class_object.namespace
       old_object.superclass = class_object.superclass
   end
  else
    @classes[class_name] = class_object
  end
end
```

By doing this, we don't need to pollute the ClassObject with `is_defined`.

File: lib/ruby_zen/indexers/iseq_indexer.rb [hunk 3]
```diff
@@ -40,7 +40,7 @@ def register_processors
         const = @engine.fetch_class(name)
         if const.nil?
           @engine.define_class(name) do
-            RubyZen::ClassObject.new(name, is_module: true)
+            RubyZen::ClassObject.new(name)
```
Comment (0x2c7) [inline lib/ruby_zen/indexers/iseq_indexer.rb]: Thanks to the class definition update, we can now be sure about the class type (module or normal class). So, I think we still need to assume that an unknown class is a module, until we have better evidence, like what Ruby is doing.

**Review by luong-komorebi — APPROVED** [review]
em thấy ok

**Review by 0x2c7 — APPROVED** [review]
LGTM 👍

**Comment by luong-komorebi** [comment]
@Khoahaha mốt anh làm chung trên repo này, đừng fork ra mỗi lần xem code checkout đồ khó khăn

**Pull request merged**